crates/edr_generic/src/eip2718.rs

```diff
@@ -0,0 +1,145 @@
+use alloy_rlp::Buf as _;
+use edr_eth::{
+    receipt::{MapReceiptLogs, Receipt, RootOrStatus},
+    transaction::TransactionType,
+    Bloom,
+};
+
+use crate::transaction;
+
+/// An compile-time typed EIP-2718 envelope for L1 Ethereum.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TypedEnvelope<DataT> {
+    /// Legacy transaction.
+    Legacy(DataT),
+    /// EIP-2930 transaction.
+    Eip2930(DataT),
+    /// EIP-1559 transaction.
+    Eip1559(DataT),
+    /// EIP-4844 transaction.
+    Eip4844(DataT),
+    /// Unrecognized transaction type.
+    Unrecognized(DataT),
+}
+
+impl<DataT> TypedEnvelope<DataT> {
+    /// Constructs a typed envelope around the given data.
+    pub fn new(data: DataT, transaction_type: transaction::Type) -> Self {
+        match transaction_type {
+            transaction::Type::Legacy => Self::Legacy(data),
+            transaction::Type::Eip2930 => Self::Eip2930(data),
+            transaction::Type::Eip1559 => Self::Eip1559(data),
+            transaction::Type::Eip4844 => Self::Eip4844(data),
+            transaction::Type::Unrecognized(_) => Self::Unrecognized(data),
+        }
+    }
+
+    /// Returns a reference to the data inside the envelope.
+    pub fn data(&self) -> &DataT {
+        match self {
+            TypedEnvelope::Legacy(data)
+            | TypedEnvelope::Eip2930(data)
+            | TypedEnvelope::Eip1559(data)
+            | TypedEnvelope::Eip4844(data)
+            | TypedEnvelope::Unrecognized(data) => data,
+        }
+    }
+
+    /// Maps the data inside the envelope to a new type.
+    pub fn map<NewDataT, F>(self, f: F) -> TypedEnvelope<NewDataT>
+    where
+        F: FnOnce(DataT) -> NewDataT,
+    {
+        match self {
+            TypedEnvelope::Legacy(data) => TypedEnvelope::Legacy(f(data)),
+            TypedEnvelope::Eip2930(data) => TypedEnvelope::Eip2930(f(data)),
+            TypedEnvelope::Eip1559(data) => TypedEnvelope::Eip1559(f(data)),
+            TypedEnvelope::Eip4844(data) => TypedEnvelope::Eip4844(f(data)),
+            TypedEnvelope::Unrecognized(data) => TypedEnvelope::Unrecognized(f(data)),
+        }
+    }
+}
+
+impl<DataT> TransactionType for TypedEnvelope<DataT> {
+    type Type = transaction::Type;
+
+    fn transaction_type(&self) -> Self::Type {
+        match self {
+            TypedEnvelope::Legacy(_) => transaction::Type::Legacy,
+            TypedEnvelope::Eip2930(_) => transaction::Type::Eip2930,
+            TypedEnvelope::Eip1559(_) => transaction::Type::Eip1559,
+            TypedEnvelope::Eip4844(_) => transaction::Type::Eip4844,
+            // TODO: Should we properly decode the transaction type?
+            TypedEnvelope::Unrecognized(_) => transaction::Type::Unrecognized(0xFF),
```

Comment:
So I'm torn here, as in practice it shouldn't matter but for example you use the library to RLP-decode the typed envelope and saturate to `transaction::Type::Legacy` here, it would not be roundtrippable per se; maybe that's okay.

Do you want me to return `transaction::Type::Legacy` with a note or `unreachable!` or something else here, specifically?

Reply:
That's correct. Internally, we never actually decoded anything that we didn't encode ourselves. That's why the guarantee holds in real use cases.

So the tricky case here is the `ChainSpecT::PooledTransaction::decode` in `pub fn handle_send_raw_transaction_request`. This should return an invalid transaction type error for any non-supported transaction types. As long as you return that for `TypedEnvelope::Unrecognized`, we should not be changing behaviour inadvertently.

Reply:
Hm, that is tricky - the `ChainSpecT::PooledTransaction` uses its decode impl but it's only compared against a `edr_eth::rlp::Error::Custom(&'static str)` and checks the string contents using a `edr_eth::transaction::Signed::is_invalid_transaction_type_error`.

While the `ProviderSpec` impl for `GenericChainSpec` has the PooledTransaction the same as edr_eth's, so we're safe, the interaction seems somewhat flimsy as we're crossing from the ChainSpec abstraction to the concrete edr_eth impls and use string checks, no less.

I'm also wary of introducing yet another type param but maybe, since we're already using generic and type-safe impls, we should introduce something like `TransactionDecodeError` and add our `InvalidType` variant on top of the RLP error or something like that? We seem to be recovering/handling this exact case often enough to uplift it to the type system but I'm curious to see what you think.

I took the liberty of matching on the const str value rather than using the if guard on the pattern in https://github.com/NomicFoundation/edr/pull/653/commits/31f123e238595cb953024f7948f88211a2abd332 for consistency - I hope you don't mind.